File: src/main/java/komu/blunt/analyzer/IdentifierRenamer.java
package komu.blunt.analyzer;

import com.google.common.collect.ImmutableList;
import komu.blunt.ast.*;
import komu.blunt.objects.Symbol;
import komu.blunt.types.patterns.*;

/**
 * Walks the AST to rename all local variables so that they become unique. This makes
 * further optimizations simpler.
 */
public final class IdentifierRenamer implements ASTVisitor<IdentifierMapping, ASTExpression>, PatternVisitor<IdentifierMapping, Pattern> {

    private int sequence = 1;

    public static ASTExpression rename(ASTExpression exp) {
        return new IdentifierRenamer().renameIdentifiers(exp, new IdentifierMapping());
    }

    public ASTExpression renameIdentifiers(ASTExpression exp, IdentifierMapping ctx) {
        return exp.accept(this, ctx);
    }

    private Pattern renameIdentifiers(Pattern pattern, IdentifierMapping ctx) {
        return pattern.accept(this, ctx);
    }
    
    private Symbol freshVariable() {
        return Symbol.symbol("$var" +  sequence++);
    }    

    @Override
    public ASTExpression visit(ASTApplication application, IdentifierMapping ctx) {
        return AST.apply(renameIdentifiers(application.func, ctx), renameIdentifiers(application.arg, ctx));
    }

    @Override
    public ASTExpression visit(ASTConstant constant, IdentifierMapping ctx) {
        return constant;
    }

    @Override
    public ASTExpression visit(ASTLambda lambda, IdentifierMapping ctx) {
        IdentifierMapping newCtx = ctx.extend();
        
        Symbol var = freshVariable();
        newCtx.put(lambda.argument, var);

        return AST.lambda(var, renameIdentifiers(lambda.body, newCtx));
    }

    @Override
    public ASTExpression visit(ASTLet let, IdentifierMapping ctx) {
        if (let.bindings.size() != 1) throw new UnsupportedOperationException();
        
        IdentifierMapping newCtx = ctx.extend();

        Symbol var = freshVariable();
        newCtx.put(let.bindings.get(0).name, var);
        
        ImplicitBinding binding = new ImplicitBinding(var, renameIdentifiers(let.bindings.get(0).expr, ctx));
        
        return AST.let(false, binding, renameIdentifiers(let.body, newCtx));
    }

    @Override
    public ASTExpression visit(ASTLetRec let, IdentifierMapping ctx) {
        if (let.bindings.size() != 1) throw new UnsupportedOperationException();

        IdentifierMapping newCtx = ctx.extend();

        Symbol var = freshVariable();
        newCtx.put(let.bindings.get(0).name, var);

        ImplicitBinding binding = new ImplicitBinding(var, renameIdentifiers(let.bindings.get(0).expr, newCtx));

        return AST.let(true, binding, renameIdentifiers(let.body, newCtx));
    }

    @Override
    public ASTExpression visit(ASTSequence sequence, IdentifierMapping ctx) {
        AST.SequenceBuilder result = AST.sequenceBuilder();
        
        for (ASTExpression exp : sequence.exps)
            result.add(renameIdentifiers(exp, ctx));
        
        return result.build();
    }

    @Override
    public ASTExpression visit(ASTSet set, IdentifierMapping ctx) {
        return AST.set(ctx.get(set.var), renameIdentifiers(set.exp, ctx));
    }

    @Override
    public ASTExpression visit(ASTVariable variable, IdentifierMapping ctx) {
        return AST.variable(ctx.get(variable.var));
    }

    @Override
    public ASTExpression visit(ASTConstructor constructor, IdentifierMapping ctx) {
        return constructor;
    }

    @Override
    public ASTExpression visit(ASTCase astCase, IdentifierMapping ctx) {
        ImmutableList.Builder<ASTAlternative> alts = ImmutableList.builder();

        for (ASTAlternative alt : astCase.alternatives) {
            IdentifierMapping newCtx = ctx.extend();
            Pattern pattern = renameIdentifiers(alt.pattern, newCtx);
            alts.add(AST.alternative(pattern, renameIdentifiers(alt.value, newCtx)));
        }

        return AST.caseExp(renameIdentifiers(astCase.exp, ctx), alts.build());
    }

    @Override
    public Pattern visit(ConstructorPattern pattern, IdentifierMapping ctx) {
        ImmutableList.Builder<Pattern> args = ImmutableList.builder();
        
        for (Pattern arg : pattern.args)
            args.add(renameIdentifiers(arg, ctx));
        
        return Pattern.constructor(pattern.name, args.build());
    }

    @Override
    public Pattern visit(LiteralPattern pattern, IdentifierMapping ctx) {
        return pattern;
    }

    @Override
    public Pattern visit(VariablePattern pattern, IdentifierMapping ctx) {
        Symbol var = freshVariable();
        ctx.put(pattern.var, var);
        return Pattern.variable(var);
    }

    @Override
    public Pattern visit(WildcardPattern pattern, IdentifierMapping ctx) {
        return pattern;
    }
}

File: src/main/kotlin/komu/blunt/analyzer/IdentifierRenamer.kt
package komu.blunt.analyzer

import komu.blunt.ast.ASTAlternative
import komu.blunt.ast.ASTExpression
import komu.blunt.ast.ImplicitBinding
import komu.blunt.objects.Symbol
import komu.blunt.types.patterns.Pattern
import komu.blunt.utils.Sequence
import java.util.*

/**
 * Walks the AST to perform α-conversion on all expressions. After performing this conversion, the
 * further optimization and analyzation passes need not know about scoping rules but can assume that
 * same name always refers to same variable.
 */
fun renameIdentifiers(exp: ASTExpression) =
    IdentifierMapping().rename(exp)

/**
 * Mapping from old identifiers to new identifiers.
 */
class IdentifierMapping(val parent: IdentifierMapping? = null) {

    private val mappings = HashMap<Symbol,Symbol>()
    private val sequence: Sequence = if (parent != null) parent.seq() else Sequence()

    private fun seq() = sequence

    operator fun get(v: Symbol): Symbol =
        mappings[v] ?: parent?.get(v) ?: v

    operator fun set(oldName: Symbol, newName: Symbol) {
        val old = mappings.put(oldName, newName)
        if (old != null)
            throw IllegalArgumentException("duplicate mapping for '$oldName'");
    }

    fun createChildContext() = IdentifierMapping(this)

    /**
     * Creates a new unique symbol based on name, installs it on the mapping and returns it.
     */
    fun freshMappingFor(name: Symbol): Symbol {
        val v = sequence.nextSymbol("\$${name}_")
        this[name] = v
        return v
    }

    fun rename(exp: ASTExpression): ASTExpression = when (exp) {
        is ASTExpression.Constant       -> exp
        is ASTExpression.Constructor    -> exp
        is ASTExpression.Sequence       -> exp.map { rename(it) }
        is ASTExpression.Application    -> exp.map { rename(it) }
        is ASTExpression.Set            -> ASTExpression.Set(this[exp.variable], rename(exp.exp))
        is ASTExpression.Variable       -> ASTExpression.Variable(this[exp.name])
        is ASTExpression.Let            -> renameLet(exp.bindings, exp.body, recursive = false)
        is ASTExpression.LetRec         -> renameLet(exp.bindings, exp.body, recursive = true)
        is ASTExpression.Lambda         -> renameLambda(exp)
        is ASTExpression.Case           -> renameCase(exp)
    }

    private fun renameLet(bindings: List<ImplicitBinding>,
                          body: ASTExpression,
                          recursive: Boolean): ASTExpression {
        val newCtx = createChildContext()
        val bindingExprCtx = if (recursive) newCtx else this

        val newBindings = bindings.map { binding ->
            val v = newCtx.freshMappingFor(binding.name)
            ImplicitBinding(v, bindingExprCtx.rename(binding.expr))
        }

        val newBody = newCtx.rename(body)

        return if (recursive)
            ASTExpression.LetRec(newBindings, newBody)
        else
            ASTExpression.Let(newBindings, newBody)
    }

    private fun renameLambda(lambda: ASTExpression.Lambda): ASTExpression {
        val newCtx = createChildContext()
        val fresh = newCtx.freshMappingFor(lambda.argument)
        return ASTExpression.Lambda(fresh, newCtx.rename(lambda.body))
    }

    private fun renameCase(case: ASTExpression.Case) =
        ASTExpression.Case(rename(case.exp), case.alternatives.map { renameAlternative(it) })

    private fun renameAlternative(alt: ASTAlternative): ASTAlternative {
        val newCtx = createChildContext()
        return ASTAlternative(newCtx.renamePattern(alt.pattern), newCtx.rename(alt.value))
    }

    private fun renamePattern(pattern: Pattern): Pattern = when (pattern) {
        is Pattern.Wildcard     -> pattern
        is Pattern.Literal      -> pattern
        is Pattern.Variable     -> Pattern.Variable(freshMappingFor(pattern.variable))
        is Pattern.Constructor  -> pattern.map { renamePattern(it) }
    }
}
